inc/zoo/swar/associative_iteration.h

```diff
@@ -1,258 +1,294 @@
+            doubling = doubling.shiftIntraLaneLeft(power, shiftClearingMask);
+        }
+        bitsToXOR >>= 1;
+        if(!bitsToXOR) { break; }
+        auto shifted = doubling.shiftIntraLaneLeft(power, shiftClearingMask);
+        doubling = doubling ^ shifted;
+        // 01...1
+        // 001...1
+        // 00001...1
+        // 000000001...1
+        shiftClearingMask =
+            shiftClearingMask & S{shiftClearingMask.value() >> power};
+        power <<= 1;
+    }
+    return S{result};
+}
+
+/// \todo because of the desirability of "accumuating" the XORs at the MSB,
+/// the parallel suffix operation is more suitable.
+template<int NB, typename B>
+constexpr SWAR<NB, B> parity(SWAR<NB, B> input) {
+    using S = SWAR<NB, B>;
+    auto preResult = parallelSuffix(input);
+    auto onlyMSB = preResult.value() & S::MostSignificantBit;
+    return S{onlyMSB};
+}
+
+
+namespace impl {
+template<int NB, typename B>
+constexpr auto makeLaneMaskFromMSB_and_LSB(SWAR<NB, B> msb, SWAR<NB, B> lsb) {
+    auto msbCopiedDown = msb - lsb;
+    auto msbReintroduced = msbCopiedDown | msb;
+    return msbReintroduced;
+}
+}
+
+template<int NB, typename B>
+constexpr auto makeLaneMaskFromLSB(SWAR<NB, B> input) {
+    using S = SWAR<NB, B>;
+    auto lsb = input & S{S::LeastSignificantBit};
+    auto lsbCopiedToMSB = S{lsb.value() << (NB - 1)};
+    return impl::makeLaneMaskFromMSB_and_LSB(lsbCopiedToMSB, lsb);
+}
+
+template<int NB, typename B>
+constexpr auto makeLaneMaskFromMSB(SWAR<NB, B> input) {
+    using S = SWAR<NB, B>;
+    auto msb = input & S{S::MostSignificantBit};
+    auto msbCopiedToLSB = S{msb.value() >> (NB - 1)};
+    return impl::makeLaneMaskFromMSB_and_LSB(msb, msbCopiedToLSB);
+}
+
+template<int NB, typename B>
+struct ArithmeticResultTriplet {
+    SWAR<NB, B> result;
+    BooleanSWAR<NB, B> carry, overflow;
+};
+
+
+/// \brief "Safe" addition, meaning non-corrupting unsigned overflow addition
+/// and producing the flags for unsigned overflow (carry) and signed overflow.
+/// This is the function to perform signed addition (that relies on supporting
+/// unsigned overflow)
+///
+/// This function is called "full addition" because it can perform the addition
+/// with all the bits of the inputs by making sure the overflow (in the
+/// unsigned sense) does not cross the lane boundary.
+/// This function has less performance than "optimistic" addition (operator+).
+/// The mechanism to manage potential overflow naturally allows the calculation
+/// of the carry and signed overflow flags for no extra performance cost.
+///
+/// The performance relies on the optimizer removing the calculation of
+/// the carry or signed overflow if they are not used.
+///
+/// When interpreted as unsigned addition, carrying out of the result is
+/// overflow.
+///
+/// The carry bit is essential to increase the precision of the results in
+/// normal arithmetic, but in unsigned SWAR it is preferable to double the
```

Comment:
doubling vs adding a bit per lane is a subtle thing that we should cover at some point

Reply:
Deferring comment improvement to get the rest of these improvements in.